### ISSUE TITLE
Security: `ignore_client_order` defaults to `false`

### ISSUE THREAD

**Issue by raphaelcohn** [ctz/rustls#23]
Shouldn't this default to `true` in `ServerSessionConfig::new()`?

Personally, I'd rather this setting didn't exist at all. Any chance we could `#cfg` this out?

**Comment by ctz** [comment]
If the server only supports acceptable-security ciphersuites then it is genuinely better to respect the client's preferences because it can express useful performance information. For example: Chrome on ARM Android puts chacha20 suites before AES-GCM because it does not have AES/GHASH hardware acceleration. Ignoring that information and negotiating AES-GCM anyway would be a shame.

There are quite a lot of guides for configuring nginx/apache-modssl which counsel for `ssl_prefer_server_ciphers on`/`SSLHonorCipherOrder on` -- this is useful and right because they configure the server to be capable of (eg.) DES or no-PFS suites but don't really want to negotiate them if it can be avoided.

**Comment by raphaelcohn** [comment]
What I'm arguing for is a sensible secure default. You've done lots of good
work in other areas. If it's a matter of secure or fast, then secure must
win.

On 10 Sep 2016 9:25 a.m., "ctz" <notifications@github.com> wrote:

If the server only supports acceptable-security ciphersuites then it is
genuinely better to respect the client's preferences because it can express
useful performance information. For example: Chrome on ARM Android puts
chacha20 suites before AES-GCM because it does not have AES/GHASH hardware
acceleration. Ignoring that information and negotiating AES-GCM anyway
would be a shame.

There are quite a lot of guides for configuring nginx/apache-modssl which
counsel for ssl_prefer_server_ciphers on/SSLHonorCipherOrder on -- this is
useful and right because they configure the server to be capable of (eg.)
DES or no-PFS suites but don't really want to negotiate them if it can be
avoided.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/ctz/rustls/issues/23#issuecomment-246099377>, or mute
the thread

**Comment by DemiMarie** [comment]
@raphaelcohn `rustls` _only_ supports secure ciphersuites.  `AES_128_GCM`, `AES_256_GCM`, and `CHACHA20_POLY1305` are _all_ top-quality ciphersuites.  ChaCha20-Poly1305 has _zero_ known security problems as long as nonces are never reused, and the only known attacks on AES-GCM (other than nonce misuse) are side-channel attacks on some (slow) non-hardware-accelerated implementations.  That makes honoring the client's wishes even more important, since we may be wasting performance and/or allowing side channel attacks otherwise.

Personally, I think that rustls should choose ChaCha20-Poly1305 if either:

- the client prefers it, or
- the client supports it, and the server doesn't support hardware acceleration for AES-GCM.

Otherwise, AES-GCM should be used.

**Comment by raphaelcohn** [comment]
@DemiMarie So, in effect, I think we're close to wanting the same thing. Using a simple boolean to express one of two strategies for choosing the ciphersuite doesn't seem quite enough. Perhaps a scheme based on a weighting for each suite? I'm not completely sure. Currently, `rustls`'s very limited set of secure ciphersuites makes this argument almost mute (and I fully support the programmers of it in making that choice), but that can and will change over time. ChaCha20-Poly1305 could become vulnerable tomorrow; AESNI instructions on some Intel chips are found to be broken, or NIST was compromised in its decision making with AES. With a SysAdmin hat on, it's far more likely that I'll be able to modify the configured ciphers. I can't imagine the average programmer exposing the preference boolean in his app's configuration.

In terms of performance, server performance **must** always trump that of the clients. At scale - something I've had more than my fair share of both programming for and administering - servers are at the limit of their load. Distributing some of that load to clients is essential. Likewise, server security is critical; break a server and all clients are usually vulnerable. Break a client, and, unless you're transmitting something inappropriate, only that client (or class of clients) is vulnerable.

Also in terms of performance, ciphersuite performance is irrelevant per se (excluding anything that would be a side channel in that statement); what matters is its perceived security. In your argument, that's hardware vs non-hardware acceleration.

Today, if I follow through your argument, you believe that we should choose ChaCha20-Poly1305 in what is in effect all circumstances unless a server doesn't support hardware acceleration. In other words, we honour the server's order. Indeed, going further, we should actually eliminate the AES-GCM suites from the server's list if we detect the absence (sic) of hardware acceleration.

So, in summary, I think a potentially more secure long term design might be:-

- Do not allow the set of ciphersuites to be configured at all;
- Instead, let a ciphersuite choice strategy be configured;
- Currently, provide just one such choice strategy that
   - Puts ChaCha20-Poly1305 top
   - Removes AES-GCM where non-hardware acceleration is detected
   - Is coded such that it can be configured using command line switches, toml, JSON, XML etc (eg an enum)
   - Possibly uses weightings, but, with such a limited cipher set, this probably isn't needed

**Comment by DemiMarie** [comment]
We still need to support AES-GCM.  Too many clients (as well as the HTTP2
protocol) require it.  But if the either the client or server prefers
ChaCha20-Poly1305, and both support it, it should be used.  It is still
very fast, and is immune to side-channel attacks.

**Comment by ctz** [comment]
If there's a time when rustls has less-good ciphersuites, work will be warranted here. I'd prefer a two level design where two lists of ciphersuites are configured:

- **Preferred**. The client's top preference that appears in this list is chosen.
- **Fallback**. The first client-supported suite in this list is chosen.

This allows configuration of many desirable behaviours.
